File: FCFS.cpp
#include <iostream>
#include <vector>
#include <iomanip>
using namespace std;
// Function to calculate Turnaround Time (TAT) and Waiting Time (WT)
void calculateTimes(const vector<int>& processes, const vector<int>& burstTime, const vector<int>& arrivalTime, vector<int>& tat, vector<int>& wt) {
    int n = processes.size();
    vector<int> completionTime(n);
    // Calculate Completion Time
    completionTime[0] = arrivalTime[0] + burstTime[0];
    for (int i = 1; i < n; i++) {
        completionTime[i] = max(completionTime[i - 1], arrivalTime[i]) + burstTime[i];
    }
    // Calculate Turnaround Time and Waiting Time
    for (int i = 0; i < n; i++) {
        tat[i] = completionTime[i] - arrivalTime[i];
        wt[i] = tat[i] - burstTime[i];
    }
}

int main() {
    int n;

    cout << "Enter the number of processes: ";
    cin >> n;

    vector<int> processes(n), burstTime(n), arrivalTime(n), tat(n), wt(n);

    cout << "Enter Arrival Time and Burst Time for each process:\n";
    for (int i = 0; i < n; i++) {
        cout << "Process " << i + 1 << ":\n";
        cout << "Arrival Time: ";
        cin >> arrivalTime[i];
        cout << "Burst Time: ";
        cin >> burstTime[i];
        processes[i] = i + 1;
    }

    // Calculate TAT and WT
    calculateTimes(processes, burstTime, arrivalTime, tat, wt);

    // Display results
    cout << "\nProcess\tArrival Time\tBurst Time\tTurnaround Time\tWaiting Time\n";
    for (int i = 0; i < n; i++) {
        cout << "P" << processes[i] << "\t\t" << arrivalTime[i] << "\t\t" << burstTime[i]
             << "\t\t" << tat[i] << "\t\t" << wt[i] << '\n';
    }

    // Calculate and display average TAT and WT
    double avgTAT = 0, avgWT = 0;
    for (int i = 0; i < n; i++) {
        avgTAT += tat[i];
        avgWT += wt[i];
    }
    avgTAT /= n;
    avgWT /= n;

    cout << "\nAverage Turnaround Time: " << fixed << setprecision(2) << avgTAT << endl;
    cout << "Average Waiting Time: " << fixed << setprecision(2) << avgWT << endl;

    return 0;
}

File: fcfs.py
def calculate_fcfs(processes, burst_time, arrival_time):
    n = len(processes)
    
    # Initialize lists for calculations
    completion_time = [0] * n
    turnaround_time = [0] * n
    waiting_time = [0] * n
    
    # Calculate Completion Time (CT)
    for i in range(n):
        if i == 0:
            completion_time[i] = arrival_time[i] + burst_time[i]
        else:
            if arrival_time[i] > completion_time[i - 1]:
                completion_time[i] = arrival_time[i] + burst_time[i]
            else:
                completion_time[i] = completion_time[i - 1] + burst_time[i]
    
    # Calculate Turnaround Time (TAT) and Waiting Time (WT)
    for i in range(n):
        turnaround_time[i] = completion_time[i] - arrival_time[i]
        waiting_time[i] = turnaround_time[i] - burst_time[i]
    
    # Display the results
    print(f"{'Process':<10}{'Arrival Time':<15}{'Burst Time':<15}{'Completion Time':<20}{'TAT':<10}{'WT':<10}")
    for i in range(n):
        print(f"{processes[i]:<10}{arrival_time[i]:<15}{burst_time[i]:<15}{completion_time[i]:<20}{turnaround_time[i]:<10}{waiting_time[i]:<10}")
    
    # Calculate and display average TAT and WT
    avg_tat = sum(turnaround_time) / n
    avg_wt = sum(waiting_time) / n
    print(f"\nAverage Turnaround Time: {avg_tat:.2f}")
    print(f"Average Waiting Time: {avg_wt:.2f}")


# Example Input
processes = ["P1", "P2", "P3", "P4"]
burst_time = [5, 3, 8, 6]
arrival_time = [0, 1, 2, 3]

# Run the FCFS Scheduling Algorithm
calculate_fcfs(processes, burst_time, arrival_time)
